src/gui/guiFormSpecMenu.cpp

```diff
@@ -4228,153 +4286,384 @@ bool GUIFormSpecMenu::OnEvent(const SEvent& event)
 					m_selected_item = new GUIInventoryList::ItemSpec(s);
+					m_selected_amount = count;
+					m_selected_dragging = button != BET_WHEEL_DOWN;
+				}
 
-					u32 count;
-					if (button == BET_RIGHT)
-						count = (s_count + 1) / 2;
-					else if (button == BET_MIDDLE)
-						count = MYMIN(s_count, 10);
-					else if (button == BET_WHEEL_DOWN)
-						count = 1;
-					else  // left
-						count = s_count;
-
-					if (!event.MouseInput.Shift) {
-						// no shift: select item
-						m_selected_amount = count;
-						m_selected_dragging = button != BET_WHEEL_DOWN;
-						m_auto_place = false;
-					} else {
-						// shift pressed: move item, right click moves 1
-						shift_move_amount = button == BET_RIGHT ? 1 : count;
+			} else if (m_selected_item) {
+				// Clicked a slot: move
+				if (button == BET_RIGHT || button == BET_WHEEL_UP)
+					move_amount = 1;
+				else if (button == BET_WHEEL_DOWN)
+					pickup_amount = MYMIN(s_count, 1);
+				else if (button == BET_MIDDLE)
+					move_amount = MYMIN(m_selected_amount, 10);
+				else if (button == BET_WHEEL_UP)
+					move_amount = 1;
+				else if (button == BET_LEFT)
+					move_amount = m_selected_amount;
+
+				if (event.MouseInput.Shift && !identical && matching) {
+					// Shift-move all of the selected item
+					move_amount = 0;
+
+					// Try to find somewhere to move the items to
+					u32 rings = m_inventory_rings.size();
+					u32 r = 0;
+					for (; r < rings; r++) {
+						const ListRingSpec &lr = m_inventory_rings[r];
+						if (lr.inventoryloc == s.inventoryloc &&
+								lr.listname == s.listname)
+							break;
 					}
-				}
-			} else { // m_selected_item != NULL
-				assert(m_selected_amount >= 1);
+					if (r >= rings) // Not found
+						break;
+					r = (r + 1) % rings;
 
-				if (s.isValid()) {
-					// Clicked a slot: move
-					if (button == BET_RIGHT || button == BET_WHEEL_UP)
-						move_amount = 1;
-					else if (button == BET_MIDDLE)
-						move_amount = MYMIN(m_selected_amount, 10);
-					else if (button == BET_LEFT)
-						move_amount = m_selected_amount;
-					// else wheeldown
+					const ListRingSpec &to_ring = m_inventory_rings[r];
+					Inventory *inv_to = m_invmgr->getInventory(to_ring.inventoryloc);
+					if (!inv_to)
+						break;
+					InventoryList *list_to = inv_to->getList(to_ring.listname);
+					if (!list_to)
+						break;
 
-					if (identical) {
-						if (button == BET_WHEEL_DOWN) {
-							if (m_selected_amount < s_count)
-								++m_selected_amount;
-						} else {
-							if (move_amount >= m_selected_amount)
-								m_selected_amount = 0;
-							else
-								m_selected_amount -= move_amount;
-							move_amount = 0;
+					ItemStack slct = list_selected->getItem(m_selected_item->i);
+
+					for (u32 i = 0; i < list_s->getSize(); i++) {
+						// Skip the selected slot
+						if (i == m_selected_item->i)
+							continue;
+						ItemStack item = list_s->getItem(i);
+
+						if (slct.name == item.name && slct.wear == item.wear &&
```

Comment:
Q: What is the goal of this code (in general)? It seems to move all similar items according to the list ring but what happens to the selected item? You previously set `move_amount = 0`, hence that one will not be moved?

Reply:
It aborts the move, and then performs a shift-move for all the matching items. The selected item stays selected at its original location.